options/ansi/generic/stdio-stubs.cpp

```diff
@@ -1003,46 +1003,69 @@ void perror(const char *string) {
 	fprintf(stderr, "%s\n", strerror(error));
 }
 
-// POSIX unlocked I/O extensions.
+// POSIX extensions.
 
-// GLIBC extensions.
+ssize_t getline(char **line, size_t *n, FILE *stream) {
+	return getdelim(line, n, '\n', stream);
+}
 
-ssize_t getline(char **line, size_t *n, FILE *file_base) {
+ssize_t getdelim(char **line, size_t *n, int delim, FILE *stream) {
 	// Otherwise, we cannot store the buffer / size.
 	if(!line || !n) {
 		errno = EINVAL;
 		return -1;
 	}
 
-	char *buffer = nullptr;
-	size_t capacity = 0;
-	if(*line) {
-		buffer = *line;
-		capacity = *n;
-	}
+	constexpr size_t chunk = 1024;
 
-	if(!capacity) {
-		buffer = reinterpret_cast<char *>(getAllocator().allocate(1024));
-		capacity = 1024;
-	}
+	char *buffer = *line;
+	size_t capacity = *n, nwritten = 0;
 
-	if(!fgets(buffer, capacity, file_base))
+	auto file = static_cast<mlibc::abstract_file *>(stream);
+	frg::unique_lock<FutexLock> lock(file->_lock);
+
+	// Avoid allocating if we've already hit the end
+	auto c = fgetc_unlocked(stream);
+	if (c == EOF || ferror(stream)) {
 		return -1;
+	} else {
+		file->unget(c);
+	}
+
+	while (true) {
+		// Fill the buffer
+		while (capacity > 0 && nwritten < capacity - 1) {
+			auto c = fgetc_unlocked(stream);
+			if (ferror(stream)) {
+				return -1;
+			} else if (c == EOF) {
+				buffer[nwritten] = 0;
+				return -1;
+			}
 
-	size_t k = strlen(buffer);
-	buffer[k] = '\n';
-	buffer[k + 1] = 0;
+			buffer[nwritten++] = c;
 
-	*line = buffer;
-	*n = capacity;
-	return k + 1;
-}
+			if (c == delim) {
+				buffer[nwritten] = 0;
+				return nwritten;
+			}
+		}
 
-ssize_t getdelim(char **line, size_t *n, int delim, FILE *file_base) {
-	__ensure(!"Not implemented");
-	__builtin_unreachable();
+		// Double the size of the buffer (but make sure it's at least 1024)
+		capacity = (capacity >= 1024) ? capacity * 2 : 1024;
+		buffer = reinterpret_cast<char *>(getAllocator().reallocate(*line, capacity));
```

Comment:
In general use ``realloc`` directly, and not frigg's allocator. Only use the latter when actually dealing with something from frigg.